### PR TITLE
Proposal: Dist shape refactor

### PR DESCRIPTION
**Pull request by twiecki**

### CODE REVIEW

File: pymc3/distributions/continuous.py [hunk 1]
```diff
@@ -194,7 +210,7 @@ def logp(self, value):
                      tau > 0, sd > 0)
 
 
-class HalfNormal(PositiveContinuous):
+class HalfNormal(PositiveUnivariateContinuous):
```
Comment (twiecki) [inline pymc3/distributions/continuous.py]: This might be a case where we want to start using multiple inheritance, e.g. `HalfNormal(Positive, Univariate, Continuous)`

File: pymc3/distributions/continuous.py [hunk 2]
```diff
@@ -1081,14 +1154,18 @@ class ExGaussian(Continuous):
         Tutorials in Quantitative Methods for Psychology,
         Vol. 4, No. 1, pp 35-45.
     """
-    def __init__(self, mu, sigma, nu, *args, **kwargs):
-        super(ExGaussian, self).__init__(*args, **kwargs)
-        self.mu = mu
-        self.sigma = sigma
-        self.nu = nu
+
+    def __init__(self, mu, sigma, nu, ndim=None, size=None, dtype=None, *args, **kwargs):
```
Comment (twiecki) [inline pymc3/distributions/continuous.py]: Not related to this PR but just seeing that we supply default values everywhere but here.

File: pymc3/distributions/discrete.py [hunk 3]
```diff
@@ -141,11 +147,14 @@ class Bernoulli(Discrete):
     p : float
         Probability of success (0 < p < 1).
     """
-    def __init__(self, p, *args, **kwargs):
-        super(Bernoulli, self).__init__(*args, **kwargs)
-        self.p = p
+    def __init__(self, p, ndim=None, size=None, dtype=None, *args, **kwargs):
+        self.p = T.as_tensor_variable(p)
```
Comment (twiecki) [inline pymc3/distributions/discrete.py]: Why is that required everywhere?

File: pymc3/distributions/distribution.py [hunk 4]
```diff
@@ -37,74 +40,211 @@ def dist(cls, *args, **kwargs):
         dist.__init__(*args, **kwargs)
         return dist
 
-    def __init__(self, shape, dtype, testval=None, defaults=[], transform=None):
-        self.shape = np.atleast_1d(shape)
-        if False in (np.floor(self.shape) == self.shape):
-            raise TypeError("Expected int elements in shape")
-        self.dtype = dtype
-        self.type = TensorType(self.dtype, self.shape)
+    def __init__(self, shape_supp, shape_ind, shape_reps, bcast, dtype, testval=None, defaults=[], transform=None):
```
Comment (twiecki) [inline pymc3/distributions/distribution.py]: Also not related to this, but mutable objects should never be `kwargs`.

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (brandonwillard) [inline pymc3/distributions/continuous.py]: Yeah, I thought so, too, but the old Java side of me tries to avoid it.

File: pymc3/distributions/continuous.py (hunk 2 above)
Comment (brandonwillard) [inline pymc3/distributions/continuous.py]: I want to say that each `Distribution` instance's constructor should supply the default value, in the same way this PR requires them to specify their support, etc.

File: pymc3/distributions/discrete.py (hunk 3 above)
Comment (brandonwillard) [inline pymc3/distributions/discrete.py]: The `T.as_tensor_variable`?  If so, I'm of the opinion that everything internal to these fundamental PyMC3 objects should only deal with Theano (where possible).  There's little reason not to--from what I can tell--and it's a simpler assumption to hold than any possible mix of Theano/not-Theano objects.
Otherwise, do you mean the `__init__` signature?  I just put those keyword args in there by default.  In a few cases, we'll probably want to make use of them.
Actually, I've even been considering adding the old `shape` param for backward compatibility.

File: pymc3/distributions/distribution.py (hunk 4 above)
Comment (brandonwillard) [inline pymc3/distributions/distribution.py]: Ah, ok, I can change that while I'm at it.

File: pymc3/distributions/discrete.py (hunk 3 above)
Comment (twiecki) [inline pymc3/distributions/discrete.py]: I meant `T.as_tensor_variable` and your reasoning makes sense, good clean living.

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (twiecki) [inline pymc3/distributions/continuous.py]: It seems pretty innocuous here.

File: pymc3/distributions/distribution.py [hunk 5]
```diff
@@ -114,10 +114,10 @@ def __init__(self, shape_supp, shape_ind, shape_reps, bcast, dtype, testval=None
             A transform function
         """
 
-        self.shape_supp = shape_supp
-        self.shape_ind = shape_ind
-        self.shape_reps = shape_reps
-        self.shape = shape_supp + shape_ind + shape_reps
+        self.shape_supp = T.cast(T.as_tensor_variable(shape_supp, ndim=1), 'int64')
```
Comment (brandonwillard) [inline pymc3/distributions/distribution.py]: `T.cast` causes problems when `T.get_vector_length` is called on these shapes.  Surprising since `T.cast` is an `Elemwise` operation that doesn't affect shape/length.  Maybe we should ask the Theano guys about this/`T.get_vector_length`?

File: pymc3/distributions/continuous.py (hunk 2 above)
Comment (brandonwillard) [inline pymc3/distributions/continuous.py]: To fix the current issues, which seem to revolve around producing a final/total shape, it might actually be ideal to already have the default value, since we can just use its shape.  Again, that would require implementations to provide the default value before calling the base `Distribution` constructor.
@twiecki, if that's a reasonable requirement, I'll just implement that now.

Comment (twiecki) [inline pymc3/distributions/continuous.py]: I'm not sure I follow what that would change on the API level, inferring the shape? If it improves the API I'm all for it.

Btw. my comment above was in terms of `mu, sigma, nu` which are args, while elsewhere we mainly use `kwargs`.

File: pymc3/distributions/distribution.py (hunk 5 above)
Comment (springcoil) [inline pymc3/distributions/distribution.py]: This is strange. Has anyone come across this yet?

Comment (brandonwillard) [inline pymc3/distributions/distribution.py]: Actually, I might've been mistaken and was referring to something like this:

```
In [20]: test_scalar = T.as_tensor_variable(1, ndim=0)

In [21]: T.get_vector_length(T.as_tensor_variable((test_scalar,)))
Out[21]: 1

In [22]: T.get_vector_length(T.cast(T.as_tensor_variable((test_scalar,)), 'int64'))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-22-0ba477b737b6> in <module>()
----> 1 T.get_vector_length(T.cast(T.as_tensor_variable((test_scalar,)), 'int64'))

/home/bwillar0/projects/python/Theano/theano/tensor/basic.pyc in get_vector_length(v)
   4282     else:
   4283         msg = str(v)
-> 4284     raise ValueError("length not known: %s" % msg)
   4285 
   4286 

ValueError: length not known: Elemwise{Cast{int64}} [id A] ''   
 |MakeVector{dtype='int8'} [id B] ''   
   |TensorConstant{1} [id C]
```

**Comment by brandonwillard** [comment]
These changes should help address #535 .

**Comment by twiecki** [comment]
@brandonwillard Are you going to make more changes (like the multiple inheritance) or want another round of review?

**Comment by brandonwillard** [comment]
I planned on spending some part of tomorrow getting the current tests to pass, so I might have a lot to push soon.

**Comment by twiecki** [comment]
Great! You'll probably also want to rebase.

**Comment by brandonwillard** [comment]
@twiecki: To really go forward with symbolic scalars in RV shapes (again, the number of scalars is **not** symbolic, since Theano doesn't handle that), it looks like we would have to put some/all of the MCMC sampling logic into a Theano `gof.Op`.  
The block-step code seems to depend on explicit shape values (within `ArrayOrdering`), and a different, non-`Op` approach to those operations doesn't immediately come to mind.  Otherwise, compiling Theano functions to return the exact shapes seems lame, but feasible. 

We could always limit the symbolic shape capabilities by assuming that the shape scalars are always constant values, then we can keep the new Theano shape logic, add some steps to extract the constant values, and work toward more Theano integration as we go.  In this case, the additions would simply serve to establish some organizational elements that are useful for multivariate work (i.e. the separate shape tuples for support, replications, etc.) and some Theano manipulations that (attempt) to perform automatic shape inference.

Any thoughts?

**Comment by twiecki** [comment]
@brandonwillard 

I think rewriting the samplers in theano would make things way more complicated. Samplers have loops and if/then clauses, both things that are quite tricky in theano. The benefit would be GPU sampling but that would be it's own project.

I think unified shape handling (even if constant) is a major improvement. Note that there are also consistencies in the univariate RVs (e.g.https://github.com/pymc-devs/pymc3/issues/1079) that hopefully can be fixed with this (is it already?).

And as you said, we can always move closer to symbolic shapes later on. Out of curiosity, would this allow changing the dimensionality of the model while sampling? Thinking of non-parametrics.

Should we take a closer look? Still needs to be rebased which I suspect will cause some hick-ups with the `T` -> `tt` change which we already have on master now.

> The latter should at least make

Seems like that stopped mid-way through.

**Comment by twiecki** [comment]
https://github.com/pymc-devs/pymc3/pull/1125/commits/4ccac8af4b9173aada161eaf577aa3fdd0eaae26 this commit doesn't seem to change anything in regards to random behavior?

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (twiecki) [inline pymc3/distributions/continuous.py]: I think we should go that route here unless it would cause other unanticipated problems.

Comment (brandonwillard) [inline pymc3/distributions/continuous.py]: Started splitting those up in the most recent commit; however, it became apparent that constructor call order, and other confusing things, could arise in this scenario.

**Comment by brandonwillard** [comment]
Yeah, that change in the commit might just be my automatic formatting.

**Comment by brandonwillard** [comment]
Regarding the samplers, we might not want/need to use the high-level Theano operations.  I was actually thinking that we move to more low-level Theano, such as implementing a `theano.Op`.  That way we might be able to put most of the current non-theano Python in a `theano.Op.perform` method.  There would be a little work in matching up the inputs and outputs, but it could be worth it in the end.

**Comment by brandonwillard** [comment]
Oh, and #1079 would be addressed by these changes (might need the associated changes to the `Distribution.random` framework, though), since it specifies an order of dimensions for all distributions (i.e. `shape_reps + shape_ind + shape_supp`).  

For the exact problem in #1079, these changes intend to allow the following:

``` python
# artificial data
k = 2
ndata = 10
data = np.array([1, 2.2, 2.5, 2.8, 4, 6, 7.2, 7.5, 7.8, 9])

# model
alpha = 0.1 * np.ones((ndata, k))
with pm.Model() as model:
    # Dirichlet will know its support is a vector of length k, i.e. shape_supp = (k,),
    # and that there are ndata-many independent copies, i.e. shape_ind = (ndata,),
    # finally, there are no replications, so shape_reps = tuple()
    p = pm.Dirichlet('p', alpha)
    # Notice below that the "shape" keyword has been exchanged for "size", and 
    # both Normal and Uniform have scalar support, i.e. shape_supp = tuple(), so 
    # we've really just specified k-many replications, i.e. shape_reps = (k,).
    # Also, shape_ind = tuple(), but would generally be whatever shape the parameters
    # are.
    mu = pm.Normal('mu', mu=5, sd=3, size=k)
    sd = pm.Uniform('sd', lower=0.1, upper=0.5, size=k)
    # Now, the following should work, since shape_supp = p.shape_supp,
    # shape_ind = p.shape_ind, shape_reps = p.shape_reps.
    categ = pm.Categorical('categ', p=p)  

    ...
```

The shapes for the last statement involving the `Categorical` distribution could be obtained directly , and in a generic way, from its parameter `p` with `shape_supp = p.shape[-1]`, `shape_ind = p.shape[:-1]`, since we know--by definition--that the Categorical distribution has vector support.
If we tried this in the current codebase, we'd `p.shape[-1]` would return a Theano scalar object and break things down the line.

File: pymc3/distributions/distribution.py [hunk 6]
```diff
@@ -37,74 +69,230 @@ def dist(cls, *args, **kwargs):
...
+            * When a distribution is multivariate
+                * and has an informative parameter, e.g. `mu`, then
+                `shape_supp = tt.shape(mu)[-ndim_supp:]` and `shape_ind = tt.shape(mu)[-ndim_supp:]`.
+        In all remaining cases the shapes must be provided by the caller.
+        `shape_reps` is always provided by the caller.
+
+
+        Parameters
+        ----------
+        shape_supp
```
Comment (twiecki) [inline pymc3/distributions/distribution.py:147]: This needs to be reformatted eventually to match the numpy-doc style.

**Comment by twiecki** [comment]
OK, I really like where this is going. The example is super helpful. Would you mind also adding one with multiple multivariate distributions? https://github.com/pymc-devs/pymc3/issues/1153 seems to be another good example.

Also want to get @fonnesbeck to take a look at the API.

So if I passed `pm.Dirichlet('p', alpha.T)` `shape_supp` would be `ndata` and `shape_ind` would be `k`? What if I did `pm.Dirichlet('p', 0.1 * np.ones(k), size=ndata)` -- would that be equivalent?

I'm not sure I fully grog the difference between `shape_ind` and `shape_rep`. For the dirichlet, couldn't `ndata` be the `shape_rep`?

What's left to do here? Certainly changing the examples (I can help with that), adding some unittests, random support (do you think this will be tricky?).

File: pymc3/distributions/distribution.py (hunk 6 above)
Comment (brandonwillard) [inline pymc3/distributions/distribution.py:147]: No problem.  Documenting is much easier with a standard!

**Comment by brandonwillard** [comment]
Yeah, in many cases `shape_ind` and `shape_reps` will serve the same purpose, but the distinction should be made.  For example, the issue I was running into with bounded/truncated sampling arose because there was no distinction.  Again, you can efficiently perform rejection sampling on many independent, identically distributed variables, but as soon as some of those variables have non-identical distributions (e.g. different means) you must handle those separately.
When someone instantiates a univariate distribution, like `Normal`, with non-scalar parameters, I'm just automatically assuming that the parameter tensors _could_ have different values, making them independent but not identical, and the `size` parameter explicitly serves to produce independent and identical copies of a distribution, since there's no way of specifying different distribution parameters with `size` alone.

To get this truly working, we need to address every current use of the shape information and either Theano-ize those parts--if possible--or restrict the symbolic scalar functionality by forcibly extracting the constant values from the shape scalars (and giving errors if they aren't effectively constant).  Similarly, we could create Theano functions that evaluate the symbolic shapes in those non-Theano places they're used.  Of course those Theano functions should not take arguments (e.g. other random variables, etc.), but perhaps they could depend on shared variables, which can be evaluated at any time I believe.
The random support is probably just a lot of typing and connecting the dots with Theano's random functions.  However, there is the matter of symbolic random values (that's what Theano's random actually returns); we might want to integrate that idea/type of object into the API.  Gotta think about that one, though.

**Comment by springcoil** [comment]
I think this needs a rebase. What else needs to be done?

**Comment by fonnesbeck** [comment]
I need to take a close look. Also, there are no tests written yet, which will be important.

**Comment by brandonwillard** [comment]
There's a lot to be done; this branch was only intended to facilitate a discussion on symbolic shape handling.  It does seem possible to make a real PR out of it, though.

In summary:
- [ ] Update the current non-symbolic uses of shape information.  This may involve further integration with Theano or on the spot evaluation of shapes.  See [my comments above](https://github.com/pymc-devs/pymc3/pull/1125#issuecomment-222574001).
  - [ ] In `ArrayOrdering`
  - [ ] In `Distribution.random()`
    - This also seems like a great time and place to start using Theano's random code and `RandomFunction` class.
  - [ ] ...
- [ ] Implement the `size` parameter for univariate distributions, so that they work like [the example described above](https://github.com/pymc-devs/pymc3/pull/1125#issuecomment-222850131).
- [ ] Finish updating multivariate distributions so that they follow [the shape convention described in the example above](https://github.com/pymc-devs/pymc3/pull/1125#issuecomment-222850131).
- [ ] Update old tests and/or (re-)add a `shape` keyword for backward compatibility.

**Comment by twiecki** [comment]
Waiting for @fonnesbeck to comment on the API.

**Comment by brandonwillard** [comment]
On a very related note, why aren't the random variable classes (i.e. `FreeRV`, `ObservedRV`, `TransformedRV`) Theano `SharedVariable`s?  
PyMC3's use of test/default values sort of implies that `SharedVariables` might be reasonable.  Also, the random variables in PyMC3 would operate more like they do in PyMC2; that is, they would have a `value` member, as well.

FYI: Theano's random framework appears to use a `gof.Op` (`RandomFunction`, specifically) for the type of object PyMC3 refers to as a random variable.  That `Op` then works like a factory to produce a `TensorType`, and update the rng state.

**Comment by fonnesbeck** [comment]
@brandonwillard you'd have to ask @jsalvatier to comment on that, as he designed most of PyMC3. PyMC3 is more functional in design than PyMC2, so I believe the idea was to have the values pass through the random variable directly to the trace, thereby relieving the random variable of having to manage its value. I can see the advantages (and disadvantages) of both approaches -- having the values together in one place (the `Point`) versus distributed across their respective variables.

**Comment by jsalvatier** [comment]
@fonnesbeck @brandonwillard  yup, that's basically the reason I didn't choose `SharedVariable`. In retrospect things have moved more in the direction of using shared variables since samplers replace some variables with shared variables. Its possible that it would good to make them be shared variables, but that would require some investigation.

**Comment by brandonwillard** [comment]
Ah, ok; I think I can see what you mean.  A random variable is expected to depend on other symbolic variables, and I'm not clear on what a "dependent" `SharedVariable` in Theano would be.

**Comment by twiecki** [comment]
@brandonwillard Sorry being unresponsive. What's the status on this?

**Comment by brandonwillard** [comment]
I'm mostly waiting for some input on the contents of [this comment](https://github.com/pymc-devs/pymc3/pull/1125#issuecomment-223839128).  I want to start some of the Theano integrations I mentioned in that comment (instead of the more conservative alternatives, like evaluating a Theano function to get non-symbolic values for shapes); however, if no one else is on board, it'll be a real waste of time.  Either way, I've got to decide on something in order to finalize some fixes for the multivariate normal and (non-i.i.d.) truncated distribution samplers.

**Comment by twiecki** [comment]
@brandonwillard I'm not totally sure I can foresee the implementation hurldes in doing the shapes in theano rather than static. You seem to suggest that it would simplify the code? In that case I would be all for it. Using theano randoms also seems like the right approach given that everything else in theano. In general, pymc3 is leveraging theano for almost everything already so a tighter integration seems to make sense and I would go as far as to suggest that the places where we instead did a homegrown alternative were poor design decisions.

Would it be a lot of time to implement the theano-shape approach for the multivariate normal case? That might most clearly show how that would look like.

What do @fonnesbeck and @jsalvatier think about doing the shapes as theano expressions?

**Comment by springcoil** [comment]
Hi all,
I'd add that I think Thomas' argument is fair from a design point of view.
But I'll defer to @fonnesbeck and @jsalvatier

On Tue, Jun 21, 2016 at 8:50 AM, Thomas Wiecki notifications@github.com
wrote:

> @brandonwillard https://github.com/brandonwillard I'm not totally sure
> I can foresee the implementation hurldes in doing the shapes in theano
> rather than static. You seem to suggest that it would simplify the code? In
> that case I would be all for it. Using theano randoms also seems like the
> right approach given that everything else in theano. In general, pymc3 is
> leveraging theano for almost everything already so a tighter integration
> seems to make sense and I would go as far as to suggest that the places
> where we instead did a homegrown alternative were poor design decisions.
> 
> Would it be a lot of time to implement the theano-shape approach for the
> multivariate normal case? That might most clearly show how that would look
> like.
> 
> What do @fonnesbeck https://github.com/fonnesbeck and @jsalvatier
> https://github.com/jsalvatier think about doing the shapes as theano
> expressions?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/pymc-devs/pymc3/pull/1125#issuecomment-227366035, or mute
> the thread
> https://github.com/notifications/unsubscribe/AA8DiGmtlbT-Jq8tXW1Ckcv_k1rO1Svcks5qN5etgaJpZM4IkNKh
> .

## 

Peadar Coyle
Skype: springcoilarch
www.twitter.com/springcoil
peadarcoyle.wordpress.com

**Comment by springcoil** [comment]
I would agree with what Thomas says. Tighter integration with theano makes
a lot of sense design wise.

**Comment by brandonwillard** [comment]
@twiecki, I don't think it would _entirely_ simplify the code, but it could simplify the use of/interaction with Theano.  It seems that drawing the line between what's symbolic and what isn't will always be an error prone chore if a very clear delineation isn't made.  I happen to prefer the line be drawn almost immediately upon entering the PyMC3 API.  User arguments are turned into Theano variables and from there everything is expected to be symbolic.
This isn't without difficulty, though; as I mentioned above, some of the `step_methods` framework relies on `FreeRV.dsize`, which becomes symbolic with the changes proposed here.
From a quick `grep`, the following files rely on `FreeRV.dsize` in a non-symbolic way:

```
blocking.py|25 col 39| slc = slice(dim, dim + var.dsize)
blocking.py|27 col 23| dim += var.dsize
model.py|527 col 21| self.dsize = tt.prod(distribution.shape)
model.py|529 col 21| self.dsize = const_dsize
step_methods/ATMCMC.py|77 col 43| self.covariance = np.eye(sum(v.dsize for v in vars))
step_methods/ATMCMC.py|97 col 49| [[v.dtype in pm.discrete_types] * (v.dsize or 1) for v in vars])
step_methods/metropolis.py|87 col 30| S = np.ones(sum(v.dsize for v in vars))
step_methods/metropolis.py|96 col 74| self.discrete = np.concatenate([[v.dtype in discrete_types ] * (v.dsize or 1) for v in vars])
step_methods/metropolis.py|251 col 25| self.dim = sum(v.dsize for v in vars)
```

It looks like `sample` would have to compile and evaluate a Theano function after refactoring some of the above operations into Theano `Op`s.

**Comment by fonnesbeck** [comment]
We could make attributes like `dsize` a property so that it returns the evaluated size when accessed, but an internal function would receive the symbolic representation.

**Comment by brandonwillard** [comment]
@fonnesbeck, yeah, that's definitely the quickest way to move this along.

**Comment by twiecki** [comment]
Sounds like a good solution to me then. @brandonwillard anything else we should decide upfront? If it can be done separately, I'd do the random() refactoring in a different PR, but I suspect they will be interconnected in which case it would make sense to have one large one.

**Comment by twiecki** [comment]
@brandonwillard just checking in if we have decided everything we need to or whether we can help some other way.

**Comment by brandonwillard** [comment]
I started prototyping a lot of my recent work in PyMC2, so I've been away from PyMC3; however, I've got a few new reasons to get back to it.  As for the direction of this shape stuff, I'll just set something up to evaluate the symbolic shapes (e.g. compile Theano functions and put those in the getters).
Will have such an update soon!

**Comment by twiecki** [comment]
OK, sounds good.

**Comment by springcoil** [comment]
Plus one to this

**Comment by twiecki** [comment]
@brandonwillard Any update on this?

**Comment by brandonwillard** [comment]
Still doing [a lot of stuff](https://github.com/brandonwillard/amimodels) in PyMC2 and the transition to PyMC3 has been too slow.
I have a few things to push, but I've run into an apparently major requirement that wasn't mentioned before: we can't just make `dsize` a property, seems like we need to pass its parents every time it's called.  Some magic with shared variables would probably do the trick, but that doesn't look necessarily easier or quicker.

**Comment by twiecki** [comment]
@brandonwillard OK, thanks for your input and thoughts. From your assessment and explorations it seems like this path isn't as straight-forward as we initially hoped it could be. What we have now is functional for the most part so I wonder if we should stick with it after all. Is that your conclusion?

**Comment by springcoil** [comment]
Any update?

**Comment by brandonwillard** [comment]
@springcoil, yes, here's a quick summary of the current state:

Motivated by some of the multivariate distribution issues, I thought it would be much easier to fix the `Distribution` class interface than attempt to hack up fixes on a nearly case-by-case basis.  #1069 is perhaps the most clear example of what the basic changes in this PR attempt to provide: a disaggregation of a `Distribution` object's dimensions that accounts for broadcasting.  Under this disaggregation `Distributions` are specified in terms of the shape of their support, the shape of the space of independent instances and the shape of the space of replications.  The "total" shape of the distribution is the concatenation of each of these spaces, i.e. `dist.shape = shape_reps + shape_ind + shape_supp`.  

The current API doesn't yield those values easily, nor does it guarantee that any method of inferring them will be consistent, which leads to the first set of fairly cosmetic changes in #1125.  Those changes mostly consist of enforcing a `Distribution` class to clearly specify each of those values upon instantiation.  For univariate distributions, this change is more boilerplate, so I attempted to automate it, but doing so was not easy (and introduced some scope creep).  

I'll follow up on the details of those difficulties, but--for now--that's the short of it.

**Comment by twiecki** [comment]
CC @AustinRochford.

File: pymc3/variational/advi.py [hunk 7]
```diff
@@ -324,17 +324,17 @@ def rvs(x):
...
     if local_RVs is not None:
         for v_, (uw, _) in local_RVs.items():
             v = get_transformed(v_)
             u = uw[0].ravel()
             w = uw[1].ravel()
-            n = r.normal(size=u.tag.test_value.shape)
+            n = r.normal(size=np.shape(u.tag.test_value))
             updates.update(
-                {v: (n * tt.exp(w) + u).reshape(v.tag.test_value.shape)})
+                {v: (n * tt.exp(w) + u).reshape(np.shape(v.tag.test_value))})
```
Comment (twiecki) [inline pymc3/variational/advi.py:337]: Why is that required?

Comment (brandonwillard) [reply pymc3/variational/advi.py:337]: `test_value` was returning a primitive (not a numpy array) in the tests.  Not sure if that's due to another issue, yet; still clearing out the big issues following the evaluation hack I just added.  Will comment on all this once the tests are passing again (or sooner, if necessary).

Comment (twiecki) [reply pymc3/variational/advi.py:337]: Not necessary, was just curious. Great to see you picking this back up!

**Comment by brandonwillard** [comment]
Quick update:  I've put in a [small hack](https://github.com/pymc-devs/pymc3/pull/1125/files#diff-eb25725602c01ec0a384e4b6d3946331R194) that should allow these `Distribution` changes to function in most cases (and, thus, provide the details necessary for addressing issues like #1069).

The basic idea is that we're fine evaluating a symbolic Theano shape object if all its inputs are constants.  We can then use the evaluated shape in all the places that aren't ready for purely symbolic shape objects.  At the very least it allows us to start setting up automatic [Theano] shape logic, which could make the high level interface simpler.

Some of this shape logic is demonstrated for the simplest case (univariate) [here](https://github.com/pymc-devs/pymc3/pull/1125/files#diff-eb25725602c01ec0a384e4b6d3946331R303).

I added a function [`has_const_inputs`](https://github.com/pymc-devs/pymc3/pull/1125/files#diff-eb25725602c01ec0a384e4b6d3946331R51) that helps determine if all inputs are constant.  Seems like a thing that might already exist in Theano, but nothing came to mind and that was simple enough to write.  The function `theano.tensor.get_scalar_constant_value` seemed like the right direction, but it doesn't handle enough cases to work for this.
Originally, the idea was to compile a Theano function to evaluate the shape object, which could also be used when the shape has variable dependencies.  The `eval` approach is just for simplicity, and, since I'm sure that some of the intermediate steps in constructing a proper Theano function could replace `has_const_inputs`, I'll revisit this Theano function approach soon.

**Comment by brandonwillard** [comment]
I gathered all the shape logic and put it into one class ([here](https://gist.github.com/brandonwillard/d4ec9f9f2753d1d9ef1aab71799860f7)).  That should cover univariate and multivariate cases, as well as keep the code organized and contained.  Next step is to go through the distributions and wire them to use that `DistributionShape` class.

I've also added some logic to handle "old style" distributions that only specify `shape`.  Hopefully that will put things back on track when everything is setup.

**Comment by twiecki** [comment]
@brandonwillard that looks like a useful abstraction. btw. will this allow dynamic shapes? CC @ferrine

**Comment by brandonwillard** [comment]
@twiecki, it won't allow dynamic shapes, but it should get rid of the need to specify `shape` in cases where the shape is actually fixed/constant.  More important, it leaves us in a position to start gradually enabling dynamic shapes.

**Comment by twiecki** [comment]
@brandonwillard That would already be a huge win.

**Comment by brandonwillard** [comment]
Looks like #2833 is doing some very related work.

**Comment by twiecki** [comment]
Yes, been meaning to reach out to you about that. Would be great to get
your input on that PR.

**Comment by brandonwillard** [comment]
Looking closer at #2833 (and its possible source/motivation, #2406), it is addressing the same thing(s).  

My first-pass general solution to the issue of symbolic shapes is [here](https://gist.github.com/brandonwillard/d4ec9f9f2753d1d9ef1aab71799860f7).  It's self-contained and requires more work in order to be used within the `Distribution` class(es), but it does single-out and provide the needed shape functionality.  Also, it might provide some shape-handling steps that can be generalized (e.g. to other backends).

In the meantime, I'm looking over the current state of the project, since I've been out of the loop for a while, and I'll try to follow-up soon.  Also, this PR could use a description summarizing its efforts!  If it isn't salvageable&mdash;relative to the current master&mdash;then perhaps I'll just submit a new PR to replace this one (and give it a description)

**Comment by twiecki** [comment]
@brandonwillard Thanks for following up. Ideally we'd come up with a joint proposal. Perhaps @aseyboldt could read through your approach and we discuss if and what we want to take. I do like the modularization of this approach.

**Comment by twiecki** [comment]
Closing because of age. Feel free to reopen if you want to fix this up.

**Comment by brandonwillard** [comment]
@twiecki [Here's a draft](https://brandonwillard.github.io/symbolic-math-in-pymc3.html) describing the kinds of things I was trying to do with PyMC3, how they relate to the issue(s) in this PR, and some proposals that could become PRs (picking up from this one, perhaps).  I'll keep updating it as I find more of my old work&mdash;which I'll most likely spread across a couple posts&mdash;but input is more than welcome at any time.

**Comment by junpenglao** [comment]
This is pretty nice @brandonwillard ! thanks for sharing.
I remembered I have seen something about combining sympy with theano to do symbolic computation - is your work in anyway related to that?

In case you missed it: we recently published our developer guide, https://docs.pymc.io/developer_guide.html, I think some graph related explanation in your blog post could go into the developer guide as well.

**Comment by twiecki** [comment]
@brandonwillard That is extremely cool! I'm excited about this deep thinking and I think it's a good time for some more low-level changes, as we're already changing a few things around for 3.7 (dropping python 2 support, sd->sigma renaming).
